include/ctype.h

```diff
@@ -22,66 +22,93 @@ extern "C" {
-extern int isdigit(int c);
+int isdigit(int c);
+#define __isdigit(c) (((c) >= '0' && (c) <= '9') ? 1 : 0)
+#define isdigit(c)   __isdigit(c)
 
 
-/* This function checks whether the passed character has graphical representation using locale. */
-extern int isgraph(int c);
+/* This function checks whether the passed character is alphanumeric. */
+int isalnum(int c);
+#define __isalnum(c) (((isalpha(c) != 0) || (isdigit(c) != 0)) ? 1 : 0)
+#define isalnum(c)   __isalnum(c)
 
 
-/* This function checks whether the passed character is lowercase letter. */
-extern int islower(int c);
+/* This function checks whether the passed character is printable. */
+int isprint(int c);
+#define __isprint(c) ((((c) > 0x1f) && ((c) < 0x7f)) ? 1 : 0)
+#define isprint(c)   __isprint(c)
 
 
-/* This function checks whether the passed character is printable. */
-extern int isprint(int c);
+/* This function checks whether the passed character has graphical representation using locale. */
+int isgraph(int c);
+#define __isgraph(c) ((((c) != ' ') && (isprint(c) != 0)) ? 1 : 0)
+#define isgraph(c)   __isgraph(c)
 
 
 /* This function checks whether the passed character is a punctuation character. */
-extern int ispunct(int c);
+int ispunct(int c);
+#define __ispunct(c) (((isgraph(c) != 0) && (isalnum(c) == 0)) ? 1 : 0)
+#define ispunct(c)   __ispunct(c)
 
 
 /* This function checks whether the passed character is white-space. */
-extern int isspace(int c);
-
-
-/* This function checks whether the passed character is an uppercase letter. */
-extern int isupper(int c);
+int isspace(int c);
+#define __isspace(c) (((c) == ' ' || (c) == '\f' || (c) == '\t' || (c) == '\n' || (c) == '\r' || (c) == '\v') ? 1 : 0)
+#define isspace(c)   __isspace(c)
 
 
 /* This function checks whether the passed character is a hexadecimal digit. */
-extern int isxdigit(int c);
+int isxdigit(int c);
+#define __isxdigit(c) (((((c) >= '0') && ((c) <= '9')) || (((c) >= 'a') && ((c) <= 'f')) || (((c) >= 'A') && ((c) <= 'F'))) ? 1 : 0)
+#define isxdigit(c)   __isxdigit(c)
 
 
 /* This function checks whether the passed character is a blank character. */
-extern int isblank(int c);
+int isblank(int c);
+#define __isblank(c) ((((c) == ' ') || ((c) == '\t')) ? 1 : 0)
+#define isblank(c)   __isblank(c)
 
 
 /* This function converts uppercase letters to lowercase. */
-extern int tolower(int c);
+int tolower(int c);
+#define __tolower(c) ((isupper(c) != 0) ? ((c) - 'A' + 'a') : c)
+#define tolower(c)   __tolower(c)
 
 
 /* This function converts lowercase letters to uppercase. */
-extern int toupper(int c);
+int toupper(int c);
+#define __toupper(c) ((islower(c) != 0) ? ((c) - 'a' + 'A') : c)
+#define toupper(c)   __toupper(c)
 
 
-static inline int isascii(int c)
-{
-	return (c & ~0x7f);
-}
+/* This function test if character is representable as ASCII value. */
+int isascii(int c);
+#define __isascii(c) ((((c) < 0) || ((c) > '\x7f')) ? 0 : 1)
```

Comment:
Why is `'\x7f'` instead of `0x7f` needed, although for iscntrl case `0x7f` is used not `'\x7f'` literal ?
```suggestion
#define __isascii(c) ((((c) < 0) || ((c) > 0x7f)) ? 0 : 1)
```